entry/src/block_component.rs

```diff
@@ -0,0 +1,2165 @@
+/// └─────────────────────────────────────────┘
+/// ```
+///
+/// ### BlockMarkerV0 Layout
+/// ```text
+/// ┌─────────────────────────────────────────┐
+/// │ Variant ID                   (1 byte)   │
+/// ├─────────────────────────────────────────┤
+/// │ Variant Data              (variable)    │
+/// └─────────────────────────────────────────┘
+/// ```
+///
+/// ### BlockMarkerV1 Layout
+/// ```text
+/// ┌─────────────────────────────────────────┐
+/// │ Variant ID                   (1 byte)   │
+/// ├─────────────────────────────────────────┤
+/// │ Variant Data              (variable)    │
+/// └─────────────────────────────────────────┘
+/// ```
+///
+/// ### BlockFooterV0 Layout
+/// ```text
+/// ┌─────────────────────────────────────────┐
+/// │ Producer Time Nanos          (8 bytes)  │
+/// ├─────────────────────────────────────────┤
+/// │ User Agent Length            (1 byte)   │
+/// ├─────────────────────────────────────────┤
+/// │ User Agent Bytes          (0-255 bytes) │
+/// └─────────────────────────────────────────┘
+/// ```
+///
+/// ### ParentReadyUpdateV0 Layout
+/// ```text
+/// ┌─────────────────────────────────────────┐
+/// │ Parent Slot                  (8 bytes)  │
+/// ├─────────────────────────────────────────┤
+/// │ Parent Block ID             (32 bytes)  │
+/// └─────────────────────────────────────────┘
+/// ```
+use {
+    crate::entry::Entry,
+    serde::{
+        de::{self, Visitor},
+        Deserialize, Deserializer, Serialize, Serializer,
+    },
+    solana_clock::Slot,
+    solana_hash::Hash,
+    std::fmt,
+};
+
+/// A block component containing either entries or special metadata.
+///
+/// `BlockComponent` enforces mutual exclusivity between entries and special data:
+/// - Can contain entries with no special data
+/// - Can contain special data with no entries
+/// - Cannot contain both entries and special data simultaneously
+/// - Empty entries with no special data is allowed (via `Default`)
+///
+/// # Serialization Format
+/// ```text
+/// ┌─────────────────────────────────────────┐
+/// │ Entry Count                  (8 bytes)  │
+/// ├─────────────────────────────────────────┤
+/// │ Entry Data OR Marker Data (variable)    │
+/// └─────────────────────────────────────────┘
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum BlockComponent {
+    Entries(Vec<Entry>),
+    BlockMarker(VersionedBlockMarker),
+}
+
+/// A versioned block marker supporting multiple format versions.
+///
+/// Provides backward compatibility through versioned variants. During deserialization,
+/// older versions are upgraded to the `Current` variant for forward compatibility.
+///
+/// # Serialization Format
+/// ```text
+/// ┌─────────────────────────────────────────┐
+/// │ Version                      (2 bytes)  │
+/// ├─────────────────────────────────────────┤
+/// │ Marker Data               (variable)    │
+/// └─────────────────────────────────────────┘
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq)]
```

Comment:
apart from BlockComponent with the actual nuance in serialize/deserialize, can we just derive Serialize / Deserialize on the remaining enum/structs? looks like the format is the same as stock serde.